File: .scatwrap/termin/kinematic/transform.py
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>termin/kinematic/transform.py</title>
</head>
<body>
<!-- BEGIN SCAT CODE -->
import math<br>
import numpy<br>
from termin.geombase import Pose3<br>
<br>
class Transform:<br>
    &quot;&quot;&quot;A class for 3D transformations tree using Pose3.&quot;&quot;&quot;<br>
<br>
    def __init__(self, local_pose, parent: 'Transform' = None, name: str = &quot;&quot;):<br>
        self._local_pose = local_pose<br>
        self.name = name<br>
        self.parent = None<br>
        self.children = []<br>
        self._global_pose = None<br>
        self._dirty = True<br>
<br>
        self._version_for_walking_to_proximal = 0<br>
        self._version_for_walking_to_distal = 0<br>
        self._version_only_my = 0<br>
<br>
        if parent:<br>
            parent.add_child(self)<br>
<br>
    def _unparent(self):<br>
        if self.parent:<br>
            self.parent.children.remove(self)<br>
            self.parent = None<br>
<br>
    def is_dirty(self):<br>
        return self._dirty<br>
<br>
    def add_child(self, child: 'Transform'):<br>
        child._unparent()<br>
        self.children.append(child)<br>
        child.parent = self<br>
        child._mark_dirty()<br>
<br>
    def link(self, child: 'Transform'):<br>
        &quot;&quot;&quot;Can be overridden to link child transforms differently.&quot;&quot;&quot;<br>
        self.add_child(child)<br>
<br>
    def relocate(self, pose: Pose3):<br>
        self._local_pose = pose<br>
        self._mark_dirty()<br>
<br>
    def relocate_global(self, global_pose):<br>
        if self.parent:<br>
            parent_global = self.parent.global_pose()<br>
            inv_parent_global = parent_global.inverse()<br>
            self._local_pose = inv_parent_global * global_pose<br>
        else:<br>
            self._local_pose = global_pose<br>
        self._mark_dirty()<br>
<br>
    def increment_version(self, version):<br>
        return (version + 1) % (2**31 - 1)<br>
<br>
    def _spread_changes_to_distal(self):<br>
        self._version_for_walking_to_proximal = self.increment_version(self._version_for_walking_to_proximal)<br>
        self._dirty = True<br>
        for child in self.children:<br>
            child._spread_changes_to_distal()<br>
<br>
    def _spread_changes_to_proximal(self):<br>
        self._version_for_walking_to_distal = self.increment_version(self._version_for_walking_to_distal)<br>
        if self.parent:<br>
            self.parent._spread_changes_to_proximal()<br>
<br>
    def _mark_dirty(self):<br>
        self._version_only_my = self.increment_version(self._version_only_my)<br>
        self._spread_changes_to_proximal()<br>
        self._spread_changes_to_distal() <br>
        # self._dirty = True # already done in _spread_changes_to_distal <br>
<br>
    def local_pose(self) -&gt; Pose3:<br>
        return self._local_pose<br>
<br>
    def global_pose(self) -&gt; Pose3:<br>
        if self._dirty:<br>
            if self.parent:<br>
                self._global_pose = self.parent.global_pose() * self._local_pose<br>
            else:<br>
                self._global_pose = self._local_pose<br>
            self._dirty = False<br>
        return self._global_pose<br>
<br>
    def _has_ancestor(self, possible_ancestor):<br>
        current = self.parent<br>
        while current:<br>
            if current == possible_ancestor:<br>
                return True<br>
            current = current.parent<br>
        return False<br>
<br>
    def set_parent(self, parent: 'Transform'):<br>
        if self._has_ancestor(parent):<br>
            raise ValueError(&quot;Cycle detected in Transform hierarchy&quot;)<br>
        self._unparent()<br>
        parent.children.append(self)<br>
        self.parent = parent<br>
        self._mark_dirty()<br>
<br>
    def transform_point(self, point: numpy.ndarray) -&gt; numpy.ndarray:<br>
        &quot;&quot;&quot;Transform a point from local to global coordinates.&quot;&quot;&quot;<br>
        global_pose = self.global_pose()<br>
        return global_pose.transform_point(point)<br>
<br>
    def transform_point_inverse(self, point: numpy.ndarray) -&gt; numpy.ndarray:<br>
        &quot;&quot;&quot;Transform a point from global to local coordinates.&quot;&quot;&quot;<br>
        global_pose = self.global_pose()<br>
        inv_global_pose = global_pose.inverse()<br>
        return inv_global_pose.transform_point(point)<br>
<br>
    def transform_vector(self, vector: numpy.ndarray) -&gt; numpy.ndarray:<br>
        &quot;&quot;&quot;Transform a vector from local to global coordinates.&quot;&quot;&quot;<br>
        global_pose = self.global_pose()<br>
        return global_pose.transform_vector(vector)<br>
<br>
    def transform_vector_inverse(self, vector: numpy.ndarray) -&gt; numpy.ndarray:<br>
        &quot;&quot;&quot;Transform a vector from global to local coordinates.&quot;&quot;&quot;<br>
        global_pose = self.global_pose()<br>
        inv_global_pose = global_pose.inverse()<br>
        return inv_global_pose.transform_vector(vector)<br>
<br>
    def __repr__(self):<br>
        return f&quot;Transform({self.name}, local_pose={self._local_pose})&quot;<br>
    <br>
    def to_trent_with_children(self, top_without_pose=False) -&gt; str:<br>
        dct = {<br>
            &quot;type&quot; : &quot;transform&quot;,<br>
            &quot;name&quot;: self.name,<br>
            &quot;children&quot;: [child.to_trent_with_children() for child in self.children]<br>
        }<br>
        if not top_without_pose:<br>
            dct[&quot;pose&quot;] = {<br>
                &quot;position&quot;: self._local_pose.lin.tolist(),<br>
                &quot;orientation&quot;: self._local_pose.ang.tolist()<br>
            }<br>
        return dct<br>
<br>
<br>
def inspect_tree(transform: 'Transform', level: int = 0, name_only: bool = False):<br>
    indent = &quot;  &quot; * level<br>
    if name_only:<br>
        print(f&quot;{indent}{transform.name}&quot;)<br>
    else:<br>
        print(f&quot;{indent}{transform}&quot;)<br>
    for child in transform.children:<br>
        inspect_tree(child, level + 1, name_only=name_only)<br>
<br>
<br>
class Transform3(Transform):<br>
    &quot;&quot;&quot;A 3D Transform with directional helpers.&quot;&quot;&quot;<br>
    def __init__(self, local_pose: Pose3 = None, parent: 'Transform3' = None, name: str = &quot;&quot;):<br>
        if local_pose is None:<br>
            local_pose = Pose3()<br>
        super().__init__(local_pose, parent, name)<br>
<br>
    def forward(self, distance: float = 1.0) -&gt; numpy.ndarray:<br>
        &quot;&quot;&quot;Get the forward direction vector in global coordinates.&quot;&quot;&quot;<br>
        local_forward = numpy.array([0.0, 0.0, distance])<br>
        return self.transform_vector(local_forward)<br>
<br>
    def up(self, distance: float = 1.0) -&gt; numpy.ndarray:<br>
        &quot;&quot;&quot;Get the up direction vector in global coordinates.&quot;&quot;&quot;<br>
        local_up = numpy.array([0.0, distance, 0.0])<br>
        return self.transform_vector(local_up)<br>
<br>
    def right(self, distance: float = 1.0) -&gt; numpy.ndarray:<br>
        &quot;&quot;&quot;Get the right direction vector in global coordinates.&quot;&quot;&quot;<br>
        local_right = numpy.array([distance, 0.0, 0.0])<br>
        return self.transform_vector(local_right)<br>
<br>
    def backward(self, distance: float = 1.0) -&gt; numpy.ndarray:<br>
        &quot;&quot;&quot;Get the backward direction vector in global coordinates.&quot;&quot;&quot;<br>
        return -self.forward(distance)<br>
<br>
    def down(self, distance: float = 1.0) -&gt; numpy.ndarray:<br>
        &quot;&quot;&quot;Get the down direction vector in global coordinates.&quot;&quot;&quot;<br>
        return -self.up(distance)<br>
<br>
    def left(self, distance: float = 1.0) -&gt; numpy.ndarray:<br>
        &quot;&quot;&quot;Get the left direction vector in global coordinates.&quot;&quot;&quot;<br>
        return -self.right(distance)<br>
<br>
<br>
<!-- END SCAT CODE -->
</body>
</html>

File: cpp/termin/bindings/geom/transform.cpp
#include "common.hpp"
#include "termin/entity/entity.hpp"

namespace termin {

// Helper to convert Python GeneralPose3 (with numpy arrays) to C++ GeneralPose3
static GeneralPose3 py_pose_to_cpp(py::object py_pose) {
    if (py_pose.is_none()) {
        return GeneralPose3::identity();
    }
    // Check if it's already a C++ GeneralPose3
    if (py::isinstance<GeneralPose3>(py_pose)) {
        return py_pose.cast<GeneralPose3>();
    }
    // Otherwise, extract ang, lin, scale from Python object
    Quat ang = Quat::identity();
    Vec3 lin = Vec3::zero();
    Vec3 scale{1.0, 1.0, 1.0};

    if (py::hasattr(py_pose, "ang")) {
        py::object ang_obj = py_pose.attr("ang");
        if (py::isinstance<Quat>(ang_obj)) {
            ang = ang_obj.cast<Quat>();
        } else {
            auto arr = ang_obj.cast<py::array_t<double>>();
            auto buf = arr.unchecked<1>();
            ang = Quat{buf(0), buf(1), buf(2), buf(3)};
        }
    }
    if (py::hasattr(py_pose, "lin")) {
        py::object lin_obj = py_pose.attr("lin");
        if (py::isinstance<Vec3>(lin_obj)) {
            lin = lin_obj.cast<Vec3>();
        } else {
            auto arr = lin_obj.cast<py::array_t<double>>();
            auto buf = arr.unchecked<1>();
            lin = Vec3{buf(0), buf(1), buf(2)};
        }
    }
    if (py::hasattr(py_pose, "scale")) {
        py::object scale_obj = py_pose.attr("scale");
        if (py::isinstance<Vec3>(scale_obj)) {
            scale = scale_obj.cast<Vec3>();
        } else {
            auto arr = scale_obj.cast<py::array_t<double>>();
            auto buf = arr.unchecked<1>();
            scale = Vec3{buf(0), buf(1), buf(2)};
        }
    }
    return GeneralPose3{ang, lin, scale};
}

void bind_transform(py::module_& m) {
    // GeneralTransform3 - thin wrapper around tc_transform*
    py::class_<GeneralTransform3>(m, "GeneralTransform3", py::dynamic_attr())
        // Default constructor - null transform
        .def(py::init<>())

        // Check validity
        .def("valid", &GeneralTransform3::valid)
        .def("__bool__", &GeneralTransform3::valid)

        // Attributes
        .def_property_readonly("name",
            [](const GeneralTransform3& self) -> py::object {
                const char* n = self.name();
                if (n) return py::str(n);
                return py::none();
            })
        .def_property_readonly("parent",
            [](const GeneralTransform3& self) -> py::object {
                GeneralTransform3 p = self.parent();
                if (!p.valid()) return py::none();
                return py::cast(p);
            })
        .def_property_readonly("children",
            [](const GeneralTransform3& self) -> py::list {
                py::list result;
                size_t count = self.children_count();
                for (size_t i = 0; i < count; i++) {
                    GeneralTransform3 child = self.child_at(i);
                    if (child.valid()) {
                        result.append(py::cast(child));
                    }
                }
                return result;
            })

        // Entity back-pointer
        .def_property_readonly("entity",
            [](const GeneralTransform3& self) -> py::object {
                Entity e = self.entity();
                if (!e.valid()) return py::none();
                return py::cast(e);
            })

        // Pose access
        .def("local_pose", &GeneralTransform3::local_pose)
        .def("global_pose", &GeneralTransform3::global_pose)
        .def("set_local_pose", [](GeneralTransform3& self, py::object pose) {
            self.set_local_pose(py_pose_to_cpp(pose));
        })
        .def("set_global_pose", [](GeneralTransform3& self, py::object pose) {
            self.set_global_pose(py_pose_to_cpp(pose));
        })

        // Relocate (accepts Python GeneralPose3, C++ GeneralPose3, Python Pose3, or C++ Pose3)
        .def("relocate", [](GeneralTransform3& self, py::object pose) {
            // Check if it's a C++ Pose3
            if (py::isinstance<Pose3>(pose)) {
                self.relocate(pose.cast<Pose3>());
            }
            // Check if it's a Python Pose3 (has ang and lin but no scale)
            else if (py::hasattr(pose, "ang") && py::hasattr(pose, "lin") && !py::hasattr(pose, "scale")) {
                py::object ang_obj = pose.attr("ang");
                py::object lin_obj = pose.attr("lin");
                Quat ang = Quat::identity();
                Vec3 lin = Vec3::zero();
                if (py::isinstance<Quat>(ang_obj)) {
                    ang = ang_obj.cast<Quat>();
                } else {
                    auto arr = ang_obj.cast<py::array_t<double>>();
                    auto buf = arr.unchecked<1>();
                    ang = Quat{buf(0), buf(1), buf(2), buf(3)};
                }
                if (py::isinstance<Vec3>(lin_obj)) {
                    lin = lin_obj.cast<Vec3>();
                } else {
                    auto arr = lin_obj.cast<py::array_t<double>>();
                    auto buf = arr.unchecked<1>();
                    lin = Vec3{buf(0), buf(1), buf(2)};
                }
                self.relocate(Pose3{ang, lin});
            } else {
                self.relocate(py_pose_to_cpp(pose));
            }
        })
        .def("relocate_global", [](GeneralTransform3& self, py::object pose) {
            // Check if it's a C++ Pose3
            if (py::isinstance<Pose3>(pose)) {
                self.relocate_global(pose.cast<Pose3>());
            }
            // Check if it's a Python Pose3 (has ang and lin but no scale)
            else if (py::hasattr(pose, "ang") && py::hasattr(pose, "lin") && !py::hasattr(pose, "scale")) {
                py::object ang_obj = pose.attr("ang");
                py::object lin_obj = pose.attr("lin");
                Quat ang = Quat::identity();
                Vec3 lin = Vec3::zero();
                if (py::isinstance<Quat>(ang_obj)) {
                    ang = ang_obj.cast<Quat>();
                } else {
                    auto arr = ang_obj.cast<py::array_t<double>>();
                    auto buf = arr.unchecked<1>();
                    ang = Quat{buf(0), buf(1), buf(2), buf(3)};
                }
                if (py::isinstance<Vec3>(lin_obj)) {
                    lin = lin_obj.cast<Vec3>();
                } else {
                    auto arr = lin_obj.cast<py::array_t<double>>();
                    auto buf = arr.unchecked<1>();
                    lin = Vec3{buf(0), buf(1), buf(2)};
                }
                self.relocate_global(Pose3{ang, lin});
            } else {
                self.relocate_global(py_pose_to_cpp(pose));
            }
        })

        // Hierarchy
        .def("add_child", [](GeneralTransform3& self, GeneralTransform3 child) {
            // add_child is implemented as set_parent on the child
            child.set_parent(self);
        })
        .def("set_parent", [](GeneralTransform3& self, py::object parent) {
            if (parent.is_none()) {
                self.unparent();
            } else {
                self.set_parent(parent.cast<GeneralTransform3>());
            }
        })
        .def("_unparent", &GeneralTransform3::unparent)
        .def("unparent", &GeneralTransform3::unparent)
        .def("link", [](GeneralTransform3& self, GeneralTransform3 child) {
            // link is implemented as set_parent on the child
            child.set_parent(self);
        })

        // Dirty tracking - removed, pool handles this internally
        // .def("is_dirty", ...) - not exposed in pool API
        // .def_property_readonly("version", ...) - not exposed in pool API

        // Transformations - return numpy arrays
        .def("transform_point", [](const GeneralTransform3& self, py::array_t<double> point) {
            Vec3 p = numpy_to_vec3(point);
            Vec3 result = self.transform_point(p);
            return vec3_to_numpy(result);
        })
        .def("transform_point_inverse", [](const GeneralTransform3& self, py::array_t<double> point) {
            Vec3 p = numpy_to_vec3(point);
            Vec3 result = self.transform_point_inverse(p);
            return vec3_to_numpy(result);
        })
        .def("transform_vector", [](const GeneralTransform3& self, py::array_t<double> vec) {
            Vec3 v = numpy_to_vec3(vec);
            Vec3 result = self.transform_vector(v);
            return vec3_to_numpy(result);
        })
        .def("transform_vector_inverse", [](const GeneralTransform3& self, py::array_t<double> vec) {
            Vec3 v = numpy_to_vec3(vec);
            Vec3 result = self.transform_vector_inverse(v);
            return vec3_to_numpy(result);
        })

        // Direction helpers - return numpy arrays
        .def("forward", [](const GeneralTransform3& self, double distance) {
            return vec3_to_numpy(self.forward(distance));
        }, py::arg("distance") = 1.0)
        .def("backward", [](const GeneralTransform3& self, double distance) {
            return vec3_to_numpy(self.backward(distance));
        }, py::arg("distance") = 1.0)
        .def("up", [](const GeneralTransform3& self, double distance) {
            return vec3_to_numpy(self.up(distance));
        }, py::arg("distance") = 1.0)
        .def("down", [](const GeneralTransform3& self, double distance) {
            return vec3_to_numpy(self.down(distance));
        }, py::arg("distance") = 1.0)
        .def("right", [](const GeneralTransform3& self, double distance) {
            return vec3_to_numpy(self.right(distance));
        }, py::arg("distance") = 1.0)
        .def("left", [](const GeneralTransform3& self, double distance) {
            return vec3_to_numpy(self.left(distance));
        }, py::arg("distance") = 1.0)

        // World matrix - return 4x4 numpy array
        .def("world_matrix", [](const GeneralTransform3& self) {
            auto result = make_mat(4, 4);
            auto buf = result.mutable_unchecked<2>();
            double m[16];
            self.world_matrix(m);
            for (int i = 0; i < 4; i++)
                for (int j = 0; j < 4; j++)
                    buf(i, j) = m[i * 4 + j];
            return result;
        })

        .def("__repr__", [](const GeneralTransform3& self) {
            const char* n = self.name();
            std::string name_str = n ? n : "<unnamed>";
            return "GeneralTransform3(" + name_str + ")";
        });
}

} // namespace termin
